### ISSUE TITLE
"Received unknown session message" when load testing

### ISSUE THREAD

**Issue by Quiark** [corda/corda#1261]
We have slightly modified the notary sample to do some load testing. We invoke around 3000 - 5000 of `DummyIssueAndSend` transactions from a single thread, using the `NotarySingle` configuration.

Usually at some point there will be hundreds of warning messages in the log:

```
[WARN ] 2017-08-16T08:43:22,142Z [Node thread] statemachine.StateMachineManager.onExistingSessionMessage - Received a session message for unknown session: SessionData(recipientSessionId=390550296595683560, payload=[[9855C494ED6B0FF9BA4A8096E45F17810B93814780528234E66FB4638EC856CE62166DA01C8AFB3450DB70829F824DACADFEE14495CCEFF5F8FC1DFAA4DD5403]]), from CN=Notary Service,O=R3,OU=corda,L=Zurich,C=CH
[WARN ] 2017-08-16T08:43:22,142Z [Node thread] statemachine.StateMachineManager.onExistingSessionMessage - Received a session message for unknown session: NormalSessionEnd(recipientSessionId=390550296595683560), from CN=Notary Service,O=R3,OU=corda,L=Zurich,C=CH
[WARN ] 2017-08-16T08:43:22,142Z [Node thread] statemachine.StateMachineManager.onExistingSessionMessage - Received a session message for unknown session: SessionConfirm(initiatorSessionId=6194288250501929307, initiatedSessionId=8967415740019770630), from CN=Notary Service,O=R3,OU=corda,L=Zurich,C=CH
[WARN ] 2017-08-16T08:43:22,143Z [Node thread] statemachine.StateMachineManager.onExistingSessionMessage - Received a session message for unknown session: SessionData(recipientSessionId=6194288250501929307, payload=[[99BD22B9492E71DBE7C4C7EC2FAE7E390C63B4D996A82CC8199F78FE45CB9394B9865B452042253A3EE677D86BBD1880DBADDF0429F8348E9A504A18BAAA4901]]), from CN=Notary Service,O=R3,OU=corda,L=Zurich,C=CH
```
I assume there is a new "session" created for each RPC request (even though it's the same connection) and that there is some kind of limit?

This configuration actually consists of the single notary node with RPC enabled, no Alice or Bob nodes are running.

**Comment by mikehearn** [comment]
Sessions only exist between peers on the network.

As you know, a couple of months ago we started a project to heavily optimise the enterprise version of the node. The developer who was doing that work is on holiday at the moment, but I think he fixed some issues that showed up under load as he went, unfortunately that work is currently in a private branch.

When @exFalso is back, I'll ask him to take a look to see if any of the bug fixes he did on his branch are related to this. Corda Open Source really is _not_ in a state where we support load testing of it at the moment unfortunately, because we've been putting all our effort into bringing the APIs to stability rather than productionisation work.

**Comment by exFalso** [comment]
These messages mean the communication between flows somehow became misaligned.

I looked at the flows involved and I'm fairly confident the issue is with the notarisation retry logic. In particular there are known issues with the node becoming extremely slow once the database becomes too big, my guess is that the growing database access time pushes notarisation latency to above 30s(!!) which is when a retry kicks in. Subsequently the original request returns, ending the flow (session). However later the second request also returns, but this time there's no session to reply to, triggering the above messages.

@adagys Do you agree the above scenario is possible? If so we need to think about how to mitigate this more gracefully

**Comment by mikehearn** [comment]
That sounds like two separate issues:

1) Single notary node mode isn't scaling.

2) Retry logic doesn't work right.

Let's keep this issue for number 2.

**Comment by adagys** [comment]
@exFalso yes that could certainly be the case here. There's a fundamental issue with the retry logic that you cannot distinguish between a slow and a crashed notary node, so I'm not sure what's the best solution here.

**Comment by adagys** [comment]
I filed a bug in Jira for this, https://r3-cev.atlassian.net/browse/CORDA-555.
@Quiark for now you can ignore these warnings, as the duplicate messages get dropped, and notarisation requests are idempotent, so this behaviour doesn't lead to any inconsistent state.

**Comment by Quiark** [comment]
Thanks. So the new information for me is that the main problem in load
testing is the H2 database. Looking forward to when it's swappable :)
